sidecar/cook/sidecar/progress.py

```diff
@@ -0,0 +1,127 @@
+        current_url = default_url
+
+        def send_progress_message(message):
+            nonlocal current_url
+            try:
+                for i in range(config.max_post_attempts):
+                    response = requests.post(current_url, allow_redirects=False, timeout=config.max_post_time_secs, json=message)
+                    if response.status_code == 202:
+                        return True
+                    elif response.is_redirect and response.status_code == 307:
+                        current_url = response.headers['location']
+                        logging.info(f'Redirected! Changed progress update callback url to: {current_url}')
+                    else:
+                        logging.warning(f'Unexpected progress update response ({response.status_code}): {response.content}')
+                        break
+                else:
+                    logging.warning(f'Reached max redirect retries ({config.max_post_redirect_follow})')
+            except Exception:
+                logging.exception(f'Error raised while posting progress update to {current_url}')
+            current_url = default_url
+            logging.info(f'Failed to post progress update. Reset progress update callback url: {current_url}')
+            return False
+
+        max_message_length = config.max_message_length
+        sample_interval_ms = config.progress_sample_interval_ms
+        sequence_counter = cst.ProgressSequenceCounter()
+        progress_updater = cst.ProgressUpdater(max_message_length, sample_interval_ms, send_progress_message)
+
+        def launch_progress_tracker(progress_location, location_tag):
+            logging.info(f'Location {progress_location} tagged as [tag={location_tag}]')
+            progress_tracker = cst.ProgressTracker(config, sequence_counter, progress_updater, progress_location, location_tag)
+            progress_tracker.start()
+            return progress_tracker
+
+        progress_locations = {config.progress_output_name: 'progress',
+                              config.stderr_file(): 'stderr',
+                              config.stdout_file(): 'stdout'}
+        logging.info(f'Progress will be tracked from {len(progress_locations)} locations')
+        progress_trackers = [launch_progress_tracker(file, name) for file, name in progress_locations.items()]
+
+        def handle_interrupt(interrupt_code, _):
+            logging.info(f'Progress Reporter interrupted with code {interrupt_code}')
+            # force send the latest progress state if available, and stop the tracker
+            for progress_tracker in progress_trackers:
+                progress_tracker.force_send_progress_update()
+                progress_tracker.stop()
+
+        signal.signal(signal.SIGINT, handle_interrupt)
+        signal.signal(signal.SIGTERM, handle_interrupt)
+
+        def dump_traceback(signal, frame):
+            faulthandler.dump_traceback()
+
+        signal.signal(signal.SIGUSR1, dump_traceback)
```

Comment:
Nitpick: I would prefer all the signal registrations to be written together.

Reply:
:heavy_check_mark: